src/gt4py/storage/storage.py

```diff
@@ -300,41 +303,48 @@ def __getitem__(self, item):
     def __setitem__(self, key, value):
         if hasattr(value, "__cuda_array_interface__"):
             gpu_view = storage_utils.gpu_view(self)
-            gpu_view[key] = cp.asarray(value.data)
-            self.device_to_host(True)
+            self.device_to_host()
+            if isinstance(value, GPUStorage):
+                value.device_to_host()
```

Comment:
I don’t understand why the two `device_to_host` calls are necessary. Why should setitem from GPU to GPU storage ever trigger a sync? The sync is only necessary on the next access on the host, right? So for me a call to `self._set_device_modified()` seems to be the right thing here, but maybe I miss something?

Reply:
Basically, if there's a stencil(s) that may still have running kernels,  the setitem may overwrite the same buffer and you get a race condition. The `device_to_host` is a misnomer in this context, all it does is to see if any of the involved storages depends on a recent kernel and waits for it to finish if necessary. Why not only one device_to_host: Previously it would just always force a synchronization, now we only do this if any of the two involved storages depends on a kernel.

Reply:
There is no race condition as long as the kernel and the copy run on the same stream, which I assume is always the case (we always use the default stream afaik).

Reply:
Note that the cupy sync in device_to_host is not a storage sync that equalizes the buffers, it is actually a `cudaDeviceSynchronize`

Reply:
I know and this is slow as it blocks everything… And as said, this is not required if everything runs on the same (default) stream.

Reply:
Sorry, our messages crossed. I see what you mean now, true, the sync would actually have to go in the "else" clause. I'll fix it.